operators/src/adapters/sparse_tiles_fill_adapter.rs

```diff
@@ -0,0 +1,696 @@
+use std::{pin::Pin, task::Poll};
+
+use futures::{ready, Stream};
+use geoengine_datatypes::{
+    primitives::TimeInterval,
+    raster::{
+        EmptyGrid2D, GeoTransform, GridBoundingBox2D, GridBounds, GridIdx2D, GridShape2D, GridStep,
+        Pixel, RasterTile2D,
+    },
+    util::Result,
+};
+use pin_project::pin_project;
+
+#[derive(Debug, PartialEq, Clone, Copy)]
+enum StateInner {
+    Initial,
+    BetweenTiles,
+    NextTile,
+    FillToEnd,
+    Ended,
+}
+
+#[derive(Debug, PartialEq, Clone)]
+struct StateOuter<T> {
+    current_idx: GridIdx2D,
+    current_time: TimeInterval,
+    next_tile: Option<RasterTile2D<T>>,
+    no_data_grid: EmptyGrid2D<T>,
+    grid_bounds: GridBoundingBox2D,
+    global_geo_transform: GeoTransform,
+    state: StateInner,
+}
+
+impl<T> StateOuter<T>
+where
+    T: Pixel,
+{
+    fn select_next_tile(&self) -> bool {
+        if let Some(t) = &self.next_tile {
+            t.tile_position == self.current_idx && t.time == self.current_time
+        } else {
+            false
+        }
+    }
+
+    pub fn poll(&self) -> bool {
```

Comment:
should this be called `should_poll`? It seems to me this conditions should be more explicitly handled by the state machine. Like if a poll_next is called check the state of the state machine depending on it call the poll. I know it's the same outcome but would make the states and their transitions more clear

Reply:
i created a new version where all transitions are in one block 467427a